source/ConditionSet.cpp

```diff
@@ -233,78 +133,215 @@ ConditionSet::ConditionSet(const DataNode &node)
 
 
 
+// Construct a terminal with a literal value;
+ConditionSet::ConditionSet(int64_t newLiteral)
+{
+	expressionOperator = OP_LIT;
+	literal = newLiteral;
+}
+
+
+
 // Load a set of conditions from the children of this node.
 void ConditionSet::Load(const DataNode &node)
 {
-	isOr = (node.Token(0) == "or");
-	if(!node.HasChildren())
-		node.PrintTrace("Error: Loading empty (sub)condition:");
-	for(const DataNode &child : node)
-		Add(child);
+	// The top-node is always an 'and' node, without the keyword.
+	expressionOperator = OP_AND;
+	ParseBooleanChildren(node);
 }
 
 
 
 // Save a set of conditions.
 void ConditionSet::Save(DataWriter &out) const
 {
-	for(const Expression &expression : expressions)
-		expression.Save(out);
+	// Default should be AND, so if it is, then just write the subsets.
+	// If this condition got optimized beyond OP_AND, then re-add the OP_AND by writing the current condition in full.
+	if(expressionOperator == OP_AND)
+		for(const auto &child : children)
+		{
+			child.SaveSubset(out);
+			out.Write();
+		}
+	else
+		SaveSubset(out);
+}
+
+
+
+void ConditionSet::SaveChild(int childNr, DataWriter &out) const
+{
+	const ConditionSet &child = children[childNr];
+	bool needBrackets = child.children.size() > 0;
+
+	if(needBrackets)
+		out.WriteToken("(");
+	children[childNr].SaveSubset(out);
+	if(needBrackets)
+		out.WriteToken(")");
+}
+
+
 
-	for(const ConditionSet &child : children)
+// Save a subset of conditions, by writing out tokens (without a newline).
+void ConditionSet::SaveSubset(DataWriter &out) const
+{
+	string opTxt = "";
+	auto it = find_if(CS_TOKEN_CONVERSION.begin(), CS_TOKEN_CONVERSION.end(),
+		[this](const std::pair<const string, ConditionSet::ExpressionOp> &e) {
+			return e.second == expressionOperator;
+		});
+	if(it != CS_TOKEN_CONVERSION.end())
+		opTxt = it->first;
+
+	switch(expressionOperator)
 	{
-		out.Write(child.isOr ? "or" : "and");
+	case OP_INVALID:
+		out.WriteToken("never");
+		break;
+	case OP_VAR:
+		out.WriteToken(conditionName);
+		break;
+	case OP_LIT:
+		out.WriteToken(literal);
+		break;
+	case OP_ADD:
+	case OP_SUB:
+	case OP_MUL:
+	case OP_DIV:
+	case OP_MOD:
+	case OP_EQ:
+	case OP_NE:
+	case OP_LE:
+	case OP_GE:
+	case OP_LT:
+	case OP_GT:
+		if(children.empty())
+		{
+			out.WriteToken("never");
+			break;
+		}
+		SaveChild(0, out);
+		for(unsigned int i = 1; i < children.size(); ++i)
+		{
+			out.WriteToken(opTxt);
+			SaveChild(i, out);
+		}
+		break;
+	case OP_AND:
+	case OP_OR:
+		out.Write(opTxt);
```

Comment:
What if these have no children?

Reply:
Then the initial parser should have changed the operator to OP_INVALID.